### DISCUSSION TITLE
How to use QuestPDF in Blazor?

### DISCUSSION THREAD

**Discussion by Alkozudo** [QuestPDF/QuestPDF#794] [Q&A] [answered]
I'm testing and playing around with QuestPDF and can easily create a PDF and let it display in the previewer in a normal console application, but it doesn't work in Blazor. I want it to display in the webpage but it doesn't. 

This is the code( from the tutorial of QuestPDF but in blazor):

@page "/pdf-generator"
@using QuestPDF.Fluent
@using QuestPDF.Helpers
@using QuestPDF.Infrastructure
@using System.Reflection.Metadata
@using testingQuestPDF.Components.Layout

<h3>PDF Generator</h3>

@code {

    public interface IDocument
    {
        DocumentMetadata GetMetadata();
        DocumentSettings GetSettings();
        void Compose(IDocumentContainer container);
    }

    public class InvoiceDocument : IDocument
    {
        public InvoiceModel Model { get; }

        public InvoiceDocument(InvoiceModel model)
        {
            Model = model;
        }

        public DocumentMetadata GetMetadata() => DocumentMetadata.Default;
        public DocumentSettings GetSettings() => DocumentSettings.Default;

        public void Compose(IDocumentContainer container)
        {
            container
                .Page(page =>
                {
                    page.Margin(50);

                    page.Header().Element(ComposeHeader);

                    page.Content().Element(ComposeContent);

                    page.Footer().AlignCenter().Text(Text =>
                    {
                        Text.CurrentPageNumber();
                        Text.Span(" / ");
                        Text.TotalPages();
                    });
                });

        
        }

        void ComposeHeader(IContainer container)
        {
            var titleStyle = TextStyle.Default.FontSize(20).SemiBold().FontColor(Colors.Blue.Medium);

            container.Row(row =>
            {
                row.RelativeItem().Column(column =>
                {
                    column.Item().Text($"Invoice #{Model.InvoiceNumber}").Style(titleStyle);

                    column.Item().Text(text =>
                    {
                        text.Span("Issue date: ").SemiBold();
                        text.Span($"{Model.IssueDate:d}");
                    });

                    column.Item().Text(text =>
                    {
                        text.Span("Due date: ").SemiBold();
                        text.Span($"{Model.DueDate:d}");
                    });
                });

                row.ConstantItem(100).Height(50).Placeholder();
            });

        }

        void ComposeContent(IContainer container)
        {
            container.PaddingVertical(40).Column(column =>
        {
            column.Spacing(5);

            column.Item().Element(ComposeTable);

            if (!string.IsNullOrWhiteSpace(Model.Comments))
                column.Item().PaddingTop(25).Element(ComposeComments);
        });
        }

        void ComposeTable(IContainer container)
        {
            container.Table(table =>
          {
              // step 1
              table.ColumnsDefinition(columns =>
              {
                  columns.ConstantColumn(25);
                  columns.RelativeColumn(3);
                  columns.RelativeColumn();
                  columns.RelativeColumn();
                  columns.RelativeColumn();
              });

              table.Header(header =>
                {
                    header.Cell().Element(CellStyle).Text("#");
                    header.Cell().Element(CellStyle).Text("Product");
                    header.Cell().Element(CellStyle).AlignRight().Text("Unit price");
                    header.Cell().Element(CellStyle).AlignRight().Text("Quantity");
                    header.Cell().Element(CellStyle).AlignRight().Text("Total");

                    static IContainer CellStyle(IContainer container)
                    {
                        return container.DefaultTextStyle(x => x.SemiBold()).PaddingVertical(5).BorderBottom(1).BorderColor(Colors.Black);
                    }
                });

              foreach (var item in Model.Items)
              {
                  table.Cell().Element(CellStyle).Text(Model.Items.IndexOf(item) + 1);
                  table.Cell().Element(CellStyle).Text(item.Name);
                  table.Cell().Element(CellStyle).AlignRight().Text($"{item.Price}$");
                  table.Cell().Element(CellStyle).AlignRight().Text(item.Quantity);
                  table.Cell().Element(CellStyle).AlignRight().Text($"{item.Price * item.Quantity}$");

                  static IContainer CellStyle(IContainer container)
                  {
                      return container.BorderBottom(1).BorderColor(Colors.Grey.Lighten2).PaddingVertical(5);
                  }
              }
          });
        }

        void ComposeComments(IContainer container)
        {
            container.Background(Colors.Grey.Lighten3).Padding(10).Column(column =>
            {
                column.Spacing(5);
                column.Item().Text("Comments").FontSize(14);
                column.Item().Text(Model.Comments);
            });
        }

        

    }

}

All it does is display the heading.
![Screenshot 2024-02-20 105342](https://github.com/QuestPDF/QuestPDF/assets/160589533/797f97d4-4328-4c14-8c92-d38b4a57bc5f)

**Comment by girlpunk** [comment, marked as answer]
In your example you've defined an interface and an implementation of that interface, but haven't requested QuestPDF generate a PDF of that, or return the PDF to the user. After creating an instance of `IDocument`, call one of the `GeneratePdf()` methods to render the PDF, then return the results of the call to your user. You may wish to use one of the overloads that accepts a stream for this.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
